### PR TITLE
Add advanced copy mechanism in Pulp3 DEB without dependency_solving

### PR DESCRIPTION
**Pull request by Manisha15**
Adds Copy mechanism to be used by Katello to copy content in CV filters

### CODE REVIEW

**Comment by quba42** [comment]
This change needs to add docs here (and possibly elsewhere): https://github.com/pulp/pulp_deb/blob/main/docs/feature_overview.rst

**Comment by quba42** [comment]
I guess this closes the following issue: https://pulp.plan.io/issues/7661

File: pulp_deb/app/tasks/copy.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        for entry in config:
+            (
+                source_repo_version,
+                dest_repo_version,
+                dest_repo,
+                content_filter,
+                dest_version_provided,
+            ) = process_entry(entry)
+
+            content_to_copy = source_repo_version.content.filter(content_filter)
```
Comment (quba42) [inline pulp_deb/app/tasks/copy.py:102]: At this point, `content_filter` is just based on the list of `pulp_href` supplied by the user. This list can include href for packages, but also href for metadata types. I have done some testing (by including silly content in the copy on purpose), and so far have not found any errors resulting from this. It does however result in nonsensical content ending up in the target repo, which could cause god knows what problems down the line. I suggest we should be a bit more opinionated here, and provide some sanity checking and guardrails for the user.

I suggest we try doing this in the `_process_config` function, by simply dropping any content of the wrong type from the users list with a warning if `structured` is set to true.

If this comment is too unclear, let's have a meeting about this.

**Review by quba42 — CHANGES_REQUESTED** [review]
Most of the requested changes are small nitpicks in docstring wording or naming.
I would make one substantive addition to the `_process_config` function.

Once that is done, this should be very much nearing completion.

File: pulp_deb/app/serializers/__init__.py [hunk 2]
```diff
@@ -23,4 +23,4 @@
 
 from .remote_serializers import AptRemoteSerializer
 
-from .repository_serializers import AptRepositorySerializer
+from .repository_serializers import AptRepositorySerializer, CopySerializer
```
[the remark this answers is not in the document]
Comment (quba42) [reply pulp_deb/app/serializers/__init__.py:26]: We decided to leave it as is.

File: CHANGES/396.feature [hunk 3]
```diff
@@ -0,0 +1 @@
+AAdd experimental advanced copy workflow for structured copying.
```
Comment (quba42) [inline CHANGES/396.feature:1]:
```suggestion
Added experimental advanced copy API with support for structured copying.
```
Noticed a typo, remembered the convention was for passed tense, and made a minor adjustment. :smile:

File: pulp_deb/app/serializers/repository_serializers.py [hunk 4]
```diff
@@ -11,3 +16,50 @@ class AptRepositorySerializer(RepositorySerializer):
...
+    """
+
+    config = serializers.JSONField(
+        help_text=_("A JSON document describing sources, destinations, and content to be copied")
+    )
+
+    structured = serializers.BooleanField(
+        help_text=_(
+            "Also copy any distributions, components, and releases as needed for any packages "
+            "being copied. This will allow for structured publications of the target repository."
```
[the remark this answers is not in the document]
Comment (Manisha15) [reply pulp_deb/app/serializers/repository_serializers.py:33]: Fixed.

File: pulp_deb/app/tasks/copy.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+from pulp_deb.app.models import AptRepository, Package, ReleaseArchitecture, PackageReleaseComponent
+
+import logging
+from gettext import gettext as _
+
+log = logging.getLogger(__name__)
+
+
+def find_structured_publish_attributes(content, src_repo_version):
```
Comment (quba42) [inline pulp_deb/app/tasks/copy.py:14]:
```suggestion
def find_structured_publish_content(content, src_repo_version):
```
Final naming nitpick.

File: pulp_deb/app/tasks/copy.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        pulp_type=PackageReleaseComponent.get_pulp_type()
+    ).only("pk")
+    architecture_ids = src_repo_version.content.filter(
+        pulp_type=ReleaseArchitecture.get_pulp_type()
+    ).only("pk")
+    package_release_components = PackageReleaseComponent.objects.filter(
+        pk__in=package_release_component_ids
+    )
+
+    children = set()
```
Comment (quba42) [inline pulp_deb/app/tasks/copy.py:35]:
```suggestion
    structured_publish_content = set()
```
If we can update all instances of `children` and `find_structured_publish_attributes` we are done!

**Review by quba42 — CHANGES_REQUESTED** [review]
Nearly done!

File: pulp_deb/app/tasks/copy.py (hunk 1 above)
Comment (quba42) [reply pulp_deb/app/tasks/copy.py:108]: Seems to work as intended.

**Comment by quba42** [comment]
I added some minimal "feature overview" documentation.
A full workflow documentation will not be added until there are pulp-cli-deb commands for this feature, and we find the time to write that documentation.

To that end: https://github.com/pulp/pulp-cli-deb/issues/22

**Pull request merged**